spark/ask_and_tell_spark.py

```diff
@@ -11,23 +11,33 @@
 from pyspark.sql import SparkSession
 
 
-def evaluate(x):
-    return (x - 2) ** 2  # Simulate a trial evaluation on a Spark worker node
+def evaluate(trial_number, x):
+    # Simulate a trial evaluation on a Spark worker node
+    # Keep the trial_number with the result as Spark does not necessarily return the results in
+    # order
+    return trial_number, (x - 2) ** 2
 
 
 if __name__ == "__main__":
     spark = SparkSession.builder.appName("OptunaSparkExample").getOrCreate()
     study = optuna.create_study()
 
-    for i in range(20):
-        trial = study.ask()
-        x = trial.suggest_float("x", -10, 10)
-        rdd = spark.sparkContext.parallelize([x])
-        result = rdd.map(evaluate).collect()[0]
-        study.tell(trial, result)
-        print(f"Trial#{trial.number}: {x=:.4e}, {result=:.4e}")
+    n_batches = 10
+    batch_size = 30
+
+    for batch in range(n_batches):
+        trials = [study.ask() for _ in range(batch_size)]
+        params = [[t.number, t.suggest_float("x", -10, 10)] for t in trials]
+        rdd = spark.sparkContext.parallelize(params, len(params))
+        results = rdd.map(lambda x: evaluate(*x)).collect()
+        for trial_number, result in results:
+            study.tell(trial_number, result)
+        print(f"Batch: {batch}")
+        print(f"\tValue: {study.best_value}")
+        print(f"\tParams: {study.best_trial.params}")
 
     print("\nBest trial:")
     print(f"\tValue: {study.best_value}")
     print(f"\tParams: {study.best_trial.params}")
+    print(f"\tNr evaluations: {len(study.trials)}")
```

Comment:
The abbreviation `Nr` may be unfamiliar to some readers (at least it was for me). How about writing it out explicitly?

```suggestion
    print(f"\tNumber of evaluations: {len(study.trials)}")
```